### PR TITLE
Add miri to rustc docs.rs page

### PR DESCRIPTION
**Pull request by ghost**
This adds miri to https://doc.rust-lang.org/nightly/nightly-rustc/

### CODE REVIEW

**Comment by InfRandomness** [comment]
I couldn't figure out a way to generate [doc.rust-lang.org/nightly/nightly-rustc](https://doc.rust-lang.org/nightly/nightly-rustc/)
locally
I tried running `/.x.py doc` but it failed, so checks might fail here too

**Comment by InfRandomness** [comment]
The pipeline doesn't seem to have failed, but since it failed on my side I don't think the same commands are being ran

**Comment by jyn514** [comment]
> The pipeline doesn't seem to have failed, but since it failed on my side I don't think the same commands are being ran

Correct, documentation is only built in bors try builds. You'll need to get x.py doc working locally before this is being merged. Since miri is a submodule, you'll need to either fix the warnings in rust-lang/miri (my preference) or allow the bootstrap warnings for miri in doc.rs. You should be able to replicate the warnings out of tree with `RUSTFLAGS="-Drustc::all -Zunstable-options" cargo doc`.

**Comment by jyn514** [comment]
> Since miri is a submodule, you'll need to either fix the warnings in rust-lang/miri (my preference

ah wait, I'm wrong - this will block adding further lints to rustdoc and rustc in the future. You need to instead make the `SourceMode` configurable through a macro parameter: https://github.com/rust-lang/rust/blob/d2598caadafb0a32c8cee58c6bcb3d80f26ee4cf/src/bootstrap/doc.rs#L710

I think `tool.rs` does something similar, you could copy from there.

**Comment by InfRandomness** [comment]
> > Since miri is a submodule, you'll need to either fix the warnings in rust-lang/miri (my preference
> 
> ah wait, I'm wrong - this will block adding further lints to rustdoc and rustc in the future. You need to instead make the `SourceMode` configurable through a macro parameter:
> 
> https://github.com/rust-lang/rust/blob/d2598caadafb0a32c8cee58c6bcb3d80f26ee4cf/src/bootstrap/doc.rs#L710
> 
> I think `tool.rs` does something similar, you could copy from there.

You mean, if there is a tool_doc!() call in tool.rs ? 
I didn't find any but I'll ripgrep and see what I can find

**Comment by InfRandomness** [comment]
> > Since miri is a submodule, you'll need to either fix the warnings in rust-lang/miri (my preference
> 
> ah wait, I'm wrong - this will block adding further lints to rustdoc and rustc in the future. You need to instead make the `SourceMode` configurable through a macro parameter:
> 
> https://github.com/rust-lang/rust/blob/d2598caadafb0a32c8cee58c6bcb3d80f26ee4cf/src/bootstrap/doc.rs#L710
> 
> I think `tool.rs` does something similar, you could copy from there.

tool.rs effectively seems to have https://github.com/rust-lang/rust/blob/d2598caadafb0a32c8cee58c6bcb3d80f26ee4cf/src/bootstrap/tool.rs#L767
but I don't really know what to do to tell the code to generate docs for that, now that i've reverted my original diff

**Comment by jyn514** [comment]
@InfRandomness the `tool.rs` file is for building miri, not documenting it. You need to change `doc.rs` to document miri.

I don't understand the rest of your question, sorry.

**Comment by InfRandomness** [comment]
> @InfRandomness the `tool.rs` file is for building miri, not documenting it. You need to change `doc.rs` to document miri.
> 
> I don't understand the rest of your question, sorry.

did you mean that I need to end up with something similar to : 
```
tool_doc!(Miri, "miri", "src/tools/miri", ["miri"], submodule=true);
```
?

**Review by jyn514 — CHANGES_REQUESTED** [review]

File: src/bootstrap/doc.rs
Comment (InfRandomness) [reply src/bootstrap/doc.rs]: Something akin to [`499ca27` (#97773)](https://github.com/rust-lang/rust/pull/97773/commits/499ca27c974b2c70db1b3a26d8b5ae8bec4607bb) ?

**Review by jyn514 — APPROVED** [review]
Thanks!

File: src/bootstrap/doc.rs
Comment (jyn514) [reply src/bootstrap/doc.rs]: Yup, exactly :)

**Comment by jyn514** [comment]
Actually, @InfRandomness can you confirm that `x.py doc src/tools/miri --stage 0` works for you after the latest commit?

**Comment by InfRandomness** [comment]
> Actually, @InfRandomness can you confirm that `x.py doc src/tools/miri --stage 0` works for you locally after the latest commit?

hm, nope, for some reason it panicked with the command you provided

**Comment by jyn514** [comment]
> I do seem to be having a couple of warnings coming from rustdocs and I would like to fix them beforehand

:+1: 

@bors r-

File: src/bootstrap/doc.rs
Comment (RalfJung) [inline src/bootstrap/doc.rs:750]: FWIW I think `in_tree: true` might be more Rust-style syntax here.

**Comment by InfRandomness** [comment]
This MR is on hold until the MIRI submodule gets updated

File: src/bootstrap/doc.rs
Comment (RalfJung) [reply src/bootstrap/doc.rs:750]: > in_tree: true feels more like when you are constructing a structure,

Yes, exactly. That is an argument in favor of `:`, isn't it?

`in_tree = true` looks like a `let`-binding, but that would mean the name is *our* choice, which it is not.

Comment (jyn514) [reply src/bootstrap/doc.rs]: This is consistent with the macro in `tool.rs`, I'd prefer to keep it as-is.
https://github.com/rust-lang/rust/blob/c62c482b82f4ea5d3d6b87d5f85b87f2fb0dc4ab/src/bootstrap/tool.rs#L693

Comment (RalfJung) [reply src/bootstrap/doc.rs]: Ah, fair. Sorry for the noise.

**Comment by RalfJung** [comment]
The Miri update landed. :)

File: src/bootstrap/doc.rs
Comment (jyn514) [reply src/bootstrap/doc.rs]: > so what do I do ? roll back

I'd appreciate that, thanks. You'll need to rebase this PR over the miri update in any case.

Comment (InfRandomness) [reply src/bootstrap/doc.rs]: > > so what do I do ? roll back
> 
> I'd appreciate that, thanks. You'll need to rebase this PR over the miri update in any case.

For some reason, git refuses to push the rebased branch

```
  1 src/bootstrap/ git … push -v --force fork refs/heads/miri-docs\:refs/heads/miri-docs
Pushing to github.com:InfRandomness/rust.git
Running pre-push script '/home/infrandomness/Documents/Dev/rust/x.py test tidy --bless'
Updating only changed submodules
Updating submodule src/tools/miri
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0        
fatal: 'fork' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: Fetched in submodule path 'src/tools/miri', but it did not contain 3361eabf3882578207a483f2cee631646e80eabc. Direct fetching of that commit failed.
Failed updating submodule. This is probably due to uncommitted local changes.
Either stash the changes by running "git stash" within the submodule's
directory, reset them by running "git reset --hard", or commit them.
To reset all submodules' changes run "git submodule foreach --recursive git reset --hard".
Build completed unsuccessfully in 0:00:00
error: failed to push some refs to 'github.com:InfRandomness/rust.git'
```

Comment (jyn514) [reply src/bootstrap/doc.rs]: This is a bug in git - try running `git submodule deinit -f src/tools/miri` and seeing if it fixes the issue.
https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/More.20submodule.20problems

Comment (InfRandomness) [reply src/bootstrap/doc.rs]: > `git submodule deinit -f src/tools/miri`

It worked but I had to remove the pre-push hook that invoked x.py because it reinitialized the submodule everytime

**Comment by InfRandomness** [comment]
Ok this is now ready to be merged

**Comment by jyn514** [comment]
> Ok this is now ready to be merged

@InfRandomness did you run `x doc --stage 0 src/tools/miri` locally to make sure it works?

**Comment by InfRandomness** [comment]
> > Ok this is now ready to be merged
> 
> @InfRandomness did you run `x doc --stage 0 src/tools/miri` locally to make sure it works?

I tried and it fails with the submodule git bug

**Comment by InfRandomness** [comment]
It started working again after cloning back up the repository, and now I have another warning I did not have before ._.

**Comment by RalfJung** [comment]
Is there a way to generate these warnings in a clone of the Miri repo? Can we have a check in Miri CI which ensures that the rustdoc check passes?

**Comment by RalfJung** [comment]
Ah, this seems to work (slightly different than the command @jyn514 shared above):
```
RUSTDOCFLAGS="-Drustdoc::all -Zunstable-options" cargo doc
```

@InfRandomness maybe a good next step would be to add that to Miri's CI.

**Comment by InfRandomness** [comment]
> Ah, this seems to work (slightly different than the command @jyn514 shared above):
> 
> ```
> 
> RUSTDOCFLAGS="-Drustdoc::all -Zunstable-options" cargo doc
> 
> ```
> 
> 
> 
> @InfRandomness maybe a good next step would be to add that to Miri's CI.

Yep, definitely, it would prevent this from happening again

**Comment by InfRandomness** [comment]
This MR will be able to land once https://github.com/rust-lang/miri/pull/2222 lands, and once the miri submodule gets updated

**Comment by jyn514** [comment]
@InfRandomness what's the status of this PR? Is it waiting on the submodule update? If not, can you rebase over the submodule update and confirm the warning has been fixed?

**Comment by InfRandomness** [comment]
> @InfRandomness what's the status of this PR? Is it waiting on the submodule update? If not, can you rebase over the submodule update and confirm the warning has been fixed?

It is waiting on the submodule update

**Comment by jyn514** [comment]
You can also update the submodule in this PR :) it's ok to do both at once.

**Comment by InfRandomness** [comment]
`./x.py doc --stage 0 src/tools/miri` seems to be running fine after updating both the submodule and rebasing on top of master

**Comment by InfRandomness** [comment]
> The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6941132216?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/7932f823b0fc912df339e28ca0ffac3a7b145e15/checks/6941132216/logs)
> Click to see the possible cause of the failure (guessed by this bot)

Huh

**Comment by jyn514** [comment]
@bors r-

@RalfJung this looks like a failure from the submodule update, not sure if you recognize it

**Comment by InfRandomness** [comment]
> @bors r-
> 
> @RalfJung this looks like a failure from the submodule update, not sure if you recognize it

I see that it said 
`We detected that this PR updated 'miri', but its tests failed.`
but I'm not sure how to reproduce that on my local machine,
do you know ?

**Comment by jyn514** [comment]
>      Running unittests src/lib.rs (build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/miri-b55e52d0cc2c0e6d)

you can probably reproduce with `x test --stage 2 src/tools/miri`. Not sure whether --stage 2 is necessary or not, --stage 1 might work.

**Comment by RalfJung** [comment]
> this looks like a failure from the submodule update, not sure if you recognize it

Yes, that's https://github.com/rust-lang/rust/issues/98107

**Comment by RalfJung** [comment]
However, this PR shouldn't have to update the submodule. Yes, Miri fails to build on master, but that happens regularly so the documentation generation code here has to be able to handle that situation.

File: src/bootstrap/doc.rs
Comment (RalfJung) [inline src/bootstrap/doc.rs:718]: If this command fails for Miri, that failure must be ignored. Miri failing to build is not a bug, it is expected. That's what the entire toolstate stuff is about.

**Comment by jyn514** [comment]
@RalfJung this has to update the submodule to pull in the documentation fix, not so it gets miri building.

Oh I see what you're saying though - if miri is broken, the docs will be broken too and we have to be able to handle that. @InfRandomness I think the right way to do that is to change https://github.com/rust-lang/rust/blob/7932f823b0fc912df339e28ca0ffac3a7b145e15/src/bootstrap/doc.rs#L732 to use `try_run` if `$in_tree` is true, and maybe print a warning if the command fails.

**Comment by RalfJung** [comment]
> this has to update the submodule to pull in the documentation fix, not so it gets miri building.

The documentation fix is already in the submodule in rustc master. The submodule update in this PR currently does not contain any doc fixes.

**Comment by RalfJung** [comment]
> to use try_run if $in_tree is true, and maybe print a warning if the command fails.

You mean `false`? Still `in_tree` seems like too innocent of a name to also control ignoring build errors...

**Comment by jyn514** [comment]
:shrug: there are only ~3 tools in tree that use submodules, I don't think it's going to cause problems in practice.

**Comment by InfRandomness** [comment]
I don't know how to resolve a conflict with a submodule,
regular 'git add' doesn't seem to work...
anyone knows?

**Comment by RalfJung** [comment]
I'd recommend to first remove the submodule update from this commit, and then rebase it.

**Comment by InfRandomness** [comment]
> I'd recommend to first remove the submodule update from this commit, and then rebase it.

With `git restore`?
that doesn't seem to work...

**Comment by RalfJung** [comment]
The submodule update is already a separate commit. So just `git reset --hard HEAD^` should do it. Then fix the submodules (`git submodule update --init --recursive`) and then rebase.

**Comment by InfRandomness** [comment]
> The submodule update is already a separate commit. So just `git reset --hard HEAD^` should do it. Then fix the submodules (`git submodule update --init --recursive`) and then rebase.

```
LANG=en_US git submodule update --init --recursive
fatal: 'fork' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: Fetched in submodule path 'src/tools/cargo', but it did not contain 4d92f07f34ba7fb7d7f207564942508f46c225d3. Direct fetching of that commit failed.
```

**Comment by RalfJung** [comment]
Looks like you got your checkout into a broken state. I remember seeing some weird behavior with submodule+rebase in the past (git is definitely terrible at handling this), but I have no idea what I did to get out of it again. Probably random bashing against various git commands until it gave in...

You also had bad luck; there was some trouble with the cargo and rust-analyzer submodules right around when you tried that (that is probably the reason of that cargo-related error).

What you want to do is get back to commit d23264b9df5bd9d10b5553956b9ae523b1125055, somehow. If it is still rebasing, try `git rebase --abort`. Then `git reset --hard d23264b9df5bd9d10b5553956b9ae523b1125055`. Then `git rebase origin/master`. And then `git submodule update --init --recursive`. Then hopefully `git status` will be clean and you can push to this PR's branch.

**Comment by InfRandomness** [comment]
I tried to git checkout the mentioned commit 
```
[infrandomness@fedora rust]$ git status
HEAD detached at d23264b9df5
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
	modified:   Cargo.lock
	modified:   src/tools/cargo (new commits)
	modified:   src/tools/miri (new commits)
	modified:   src/tools/rust-analyzer (new commits, modified content)

no changes added to commit (use "git add" and/or "git commit -a")
[infrandomness@fedora rust]$ git rebase --abort
fatal: No rebase in progress?
```

**Comment by InfRandomness** [comment]
Ok I've had enough of this git bullcrap

I'm scraping all of this with my fork, recreating a new fork and a new PR.

**Comment by InfRandomness** [comment]
Nvm I won't even bother opening a new PR
I'm done with this.

**Comment by RalfJung** [comment]
Ah, one submodule has "modified content". Not sure if you edited it or some script changed a file in there, but I don't know of a good way to tell git to reset all submodule *contents*, you have to go to each affected submodule and do it by hand. :(

I re-submitted this as https://github.com/rust-lang/rust/pull/98714.

**Pull request closed without merging**
